### PR TITLE
Detect undocumented protected classes, properties, and functions

### LINKED ISSUE #4633: Rules for undocumented protected members
The rule or changes you're looking for might have already been suggested!
Please search in the [issues](https://github.com/detekt/detekt/issues) before creating a new one.

## Expected Behavior of the rule

Would like to get a warning when I have protected functions, properties, or classes that are not documented.  

## Context

As a library provider with classes that are intended to be extended by consuming code, my protected members are part of the public API and should get documented.  I could see this as being an optional flag on the public rules that already exist.

### PR DESCRIPTION
**Pull request by KengoTODA**
Hello :wave: thanks for making such a great product open and shared!
My team is using Detekt and I want to learn more about, so today I want to suggest a PR that can close #4633

I simply follow the discussion in #4633, so added a configuration to each existing rule.
Honestly speaking, I have less confidence in the naming of configurations, hope that we will have a discussion about it in this PR.

### CODE REVIEW

File: detekt-core/src/main/resources/default-detekt-config.yml [hunk 1]
```diff
@@ -78,12 +78,15 @@ comments:
     searchInInnerClass: true
     searchInInnerObject: true
     searchInInnerInterface: true
+    searchProtectedClass: false
```
Comment (cortinico) [inline detekt-core/src/main/resources/default-detekt-config.yml:81]: To be consistent, I would rename this to `searchInProtectedClass`

Comment (KengoTODA) [reply detekt-core/src/main/resources/default-detekt-config.yml]: Renamed in the commit 6e2f19b, thank you!

File: detekt-psi-utils/api/detekt-psi-utils.api [hunk 2]
```diff
@@ -133,7 +134,8 @@ public final class io/gitlab/arturbosch/detekt/rules/ThrowExtensionsKt {
 }
 
 public final class io/gitlab/arturbosch/detekt/rules/TraversingKt {
-	public static final fun isPublicInherited (Lorg/jetbrains/kotlin/psi/KtNamedDeclaration;)Z
+	public static final fun isPublicInherited (Lorg/jetbrains/kotlin/psi/KtNamedDeclaration;Z)Z
+	public static synthetic fun isPublicInherited$default (Lorg/jetbrains/kotlin/psi/KtNamedDeclaration;ZILjava/lang/Object;)Z
```
Comment (BraisGabin) [inline detekt-psi-utils/api/detekt-psi-utils.api:138]: same

File: detekt-psi-utils/api/detekt-psi-utils.api [hunk 3]
```diff
@@ -105,7 +105,8 @@ public final class io/gitlab/arturbosch/detekt/rules/KtModifierListKt {
 	public static final fun isOperator (Lorg/jetbrains/kotlin/psi/KtModifierListOwner;)Z
 	public static final fun isOverride (Lorg/jetbrains/kotlin/psi/KtModifierListOwner;)Z
 	public static final fun isProtected (Lorg/jetbrains/kotlin/psi/KtModifierListOwner;)Z
-	public static final fun isPublicNotOverridden (Lorg/jetbrains/kotlin/psi/KtModifierListOwner;)Z
+	public static final fun isPublicNotOverridden (Lorg/jetbrains/kotlin/psi/KtModifierListOwner;Z)Z
+	public static synthetic fun isPublicNotOverridden$default (Lorg/jetbrains/kotlin/psi/KtModifierListOwner;ZILjava/lang/Object;)Z
```
[the remark this answers is not in the document]
Comment (KengoTODA) [reply detekt-psi-utils/api/detekt-psi-utils.api]: Great! Really thank you for telling this point to me, I didn't notice this important point before.

**Review by BraisGabin — APPROVED** [review]

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt [hunk 4]
```diff
@@ -44,6 +44,9 @@ class UndocumentedPublicClass(config: Config = Config.empty) : Rule(config) {
     @Configuration("if inner interfaces should be searched")
     private val searchInInnerInterface: Boolean by config(true)
 
+    @Configuration("if protected class should be searched")
```
Comment (schalkms) [inline detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:47]:
```suggestion
    @Configuration("if protected classes should be searched")
```

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicFunction.kt [hunk 5]
```diff
@@ -27,6 +30,9 @@ class UndocumentedPublicFunction(config: Config = Config.empty) : Rule(config) {
         Debt.TWENTY_MINS
     )
 
+    @Configuration("if protected function should be searched")
```
Comment (schalkms) [inline detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicFunction.kt:33]:
```suggestion
    @Configuration("if protected functions should be searched")
```

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicProperty.kt [hunk 6]
```diff
@@ -31,6 +33,9 @@ class UndocumentedPublicProperty(config: Config = Config.empty) : Rule(config) {
         Debt.TWENTY_MINS
     )
 
+    @Configuration("if protected function should be searched")
```
Comment (schalkms) [inline detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicProperty.kt:36]:
```suggestion
    @Configuration("if protected functions should be searched")
```

**Review by schalkms — APPROVED** [review]
Thanks for this nice PR! 👍

**Review by cortinico — APPROVED** [review]

**Comment by KengoTODA** [comment]
Thank you!

**Pull request merged**